File: C++/day1.cpp
/*
Advent of Code 2020 - Day 1

This code will search through a list of numbers and return the product of
the two or three that add to 2020

Tom Kite - 01/12/20
*/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

vector<int> read_data(string filename)
{
    ifstream ifile;
    ifile.open(filename);

    if (!ifile) { throw; }

    vector<int> data{};
    string line;
    while (getline(ifile, line))
    {
        data.push_back(stoi(line));
    }
    ifile.close();
    return data;
}

int find_2_sum(vector<int> data, int total=2020)
{
    for (int i{0}; i<data.size(); i++)
    {
        for (int j{i+1}; j<data.size(); j++)
        {
            if (data[i] + data[j] == total){ return data[i] * data[j]; }
        }
    }
    return -1;
}

int find_3_sum(vector<int> data, int total=2020)
{
    for (int i{0}; i<data.size(); i++)
    {
        for (int j{i+1}; j<data.size(); j++)
        {
            if (data[i] + data[j] >= total) { continue; }
            for (int k{j+1}; k<data.size(); k++)
            {
                if (data[i] + data[j] + data[k] == total)
                {
                    return data[i] * data[j] * data[k];
                }
            }
        }
    }
    return -1;
}

int main()
{
    vector<int> data;
    try { data = read_data("../data/day1.dat"); }
    catch (...) { cout << "Could not open file!" << endl; return 1; }

    cout << find_2_sum(data) << endl;
    cout << find_3_sum(data) << endl;

    return 0;
}

File: python/headers/day1.py
# -*- coding: utf-8 -*-
"""
Day 1 of Advent of Code 2020

This script will find 2 (3) entries in a list of numbers that add to 2020,
and then return the multiplication of the 2 (3)

Tom Kite - 01/12/2020
"""

from aoc_tools.advent_timer import Advent_Timer
import numpy as np


def part1(filename):
    data = np.genfromtxt(filename, dtype='int')

    for i, num_a in enumerate(data):
        for num_b in data[i+1:]:
            if num_a + num_b == 2020:
                print("The numbers are {} and {}.\n"
                      "Giving a product of {}."
                      .format(num_a, num_b, num_a*num_b))
                return


def part2(filename):
    data = np.genfromtxt(filename, dtype='int')

    for i, num_a in enumerate(data):
        for j, num_b in enumerate(data[i+1:]):
            if num_a + num_b >= 2020:
                continue
            for num_c in data[j+1:]:
                if num_a + num_b + num_c == 2020:
                    print("The numbers are {}, {} and {}.\n"
                          "Giving a product of {}."
                          .format(num_a, num_b, num_c, num_a*num_b*num_c))
                    return


if __name__ == "__main__":
    timer = Advent_Timer()

    print("Part 1:")
    part1("../../data/day1.dat")
    timer.checkpoint_hit()

    print("\nPart 2:")
    part2("../../data/day1.dat")
    timer.checkpoint_hit()

    timer.end_hit()
